app/src/main/java/com/example/habitsmasher/UserRegistrationFragment.java

```diff
@@ -89,12 +98,45 @@ public void onClick(View view) {
                     return;
                 }
 
-                _progressBar.setVisibility(View.VISIBLE);
+                checkForUniqueUsername(isUsernameTaken -> {
+                    if (isUsernameTaken) {
+                        Log.d(TAG, "This username already exists");
+
+                        usernameInput.setError(THIS_USERNAME_IS_ALREADY_TAKEN_MESSAGE);
+                        usernameInput.requestFocus();
+
+                        _progressBar.setVisibility(View.INVISIBLE);
+                    } else {
+                        Log.d(TAG, "Username is unique!");
+
+                        createNewUserWithEmailAndPassword(validator.getValidEmailForSignUp(),
+                                                          validator.getValidPasswordForSignUp(),
+                                                          validator.getValidUsernameForSignUp());
+                    }
+                }, usernameInput.getText().toString().trim());
+            }
+        });
+    }
+
+    private void checkForUniqueUsername(FirestoreCallback firestoreCallback, String username) {
+        Log.d(TAG, "Fetching users");
+
+        FirebaseFirestore db = FirebaseFirestore.getInstance();
+        CollectionReference usersCollectionRef = db.collection(USERS_COLLECTION_PATH);
 
-                createNewUserWithEmailAndPassword(validator.getValidEmailForSignUp(),
-                                                  validator.getValidPasswordForSignUp(),
-                                                  validator.getValidUsernameForSignUp());
+        usersCollectionRef.get().addOnSuccessListener(queryDocumentSnapshots -> {
+            for (DocumentSnapshot document: queryDocumentSnapshots.getDocuments()) {
+                if (document.get(USERNAME_FIELD).toString().equals(username)) {
+                    Log.d(TAG, "Username taken: " + document.get(USERNAME_FIELD).toString());
+
+                    firestoreCallback.onCallback(true);  // if username is taken, set the flag
+                    return;
+                } else {
+                    Log.d(TAG, "Username does not equal: " + document.get(USERNAME_FIELD).toString());
+                }
             }
+            Log.d(TAG, "Finished fetching users, username does not exist");
+            firestoreCallback.onCallback(false);  // if username is not taken, set the flag to false
```

Comment:
This chunk of code iterates through all the users, checking the `username` field to make sure it's unique. If the username is taken, we set the flag in the `onCallback` method of the `FirestoreCallback` to true and **return**, otherwise it is set to false at the end of the method.